app/scripts/services/planner.service.js

```diff
@@ -66,18 +66,41 @@
             // check to see if we've finished all the bosses
             if (step.bosses) {
                 angular.forEach(step.bosses, function(boss) {
-                    if (boss.isAlliance && items.isAlliance && items.lookup[boss.ID] === undefined) {
-                        neededBosses.push(boss);
-                        completed = false;
+
+                    var bossIsNeutral = !boss.isAlliance && !boss.isHorde,
+                        character = items, // aliasing for clarity
+                        characterNeedsBoss = function(id){ return !character.lookup[id]; },
+                        addBoss = function(boss) {
+                            neededBosses.push(boss);
+                            completed = false;
+                        };
+
+                    if (showAll) { addBoss(boss); return; }
+                    if (boss.ID === undefined) { return; } // continue the loop, bad boss data
```

Comment:
Wouldn't it be simpler to just add below that:

```
if (!characterNeedsBoss(boss.ID)) { return; }
```

and then never do this check again?

Reply:
That's a great suggestion @seirl good eye. There's always something lost when moving from pseudo-code to real code, in this case, the optimization that was staring me in the face.

Makes it a lot more readable, thanks